src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala

```diff
@@ -349,7 +349,17 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
 
         override def withOuterTest(orig: Tree)(testedBinder: Symbol, expectedTp: Type): Tree = {
           val expectedPrefix = expectedTp.prefix
-          val testedPrefix = testedBinder.info.prefix
+          def prefixOf(tp: Type): Type = tp.normalize match {
+            case RefinedType(parents, _) =>
+              val temp1 = parents.filterNot(_.typeSymbol.isStatic)
+              val temp2 = temp1.map(sym => prefixOf(sym))
+              temp2.distinct match {
+                case prefix :: Nil => prefix
+                case _ => NoType
+              }
+            case tp => tp.prefix
+          }
+          val testedPrefix = prefixOf(abstractTypesToBounds(testedBinder.info))
```

Comment:
Good news 🎉  (IMHO) I rebased this on 2.13 (where fixes have been landing over the last 18 months) to study it and found that the whole `RefinedType` shenanigans above (as well as normalizing) isn't required.  The following seems sufficient to get TermName usage to stop warning (i.e. the pos test passes):

```diff
-          val expectedPrefix = expectedTp.prefix
-          val testedPrefix = testedBinder.info.prefix
+          val expectedPrefix   = expectedTp.prefix
+          val testedTp         = testedBinder.info
+          val testedTpToBounds = abstractTypesToBounds(testedTp)
+          val testedPrefix     = testedTpToBounds.prefix
```

While I'm here, here are the values and what subclass they are:

```
expectedTp           = Test1.g.TermName     :        ClassNoArgsTypeRef
expectedPrefix       = Test1.g.type         :          UniqueSingleType
testedBinder         = value x1             :                TermSymbol
testedTp             = sym.NameType         :     AbstractNoArgsTypeRef
testedTpToBounds     = Test1.g.Name         :        ClassNoArgsTypeRef
testedPrefix         = Test1.g.type         :          UniqueSingleType
```

So then it's a question of whether it's sane/principled/theory to throw in a little `abstractTypesToBounds` type mapping here...  Looks like pre-existing usages of that type map are all related to inferring types in patterns (`inferTypedPattern` and `inferConstructorInstance`) which is some form of "precedent".  Collapsed it makes this a 1 line change, too.

Reply:
I think the `RefinedType` shenanigans were pushed to satisfy existing tests:
> Pushed another version to try to walk the tightrope between pos/t11534 and run/reflect-priv-ctor.scala. Marking as WIP as this isn't super well thought through...

Did you check those too?

Reply:
`pos/t11534` is the test I was studying but, you're right, my reduction causes `abstractTypesToBounds` to induce a new outer reference warning, by finding the "wrong" prefix:

```
expectedTp           = reflect.runtime.universe.MethodSymbol                                                                                        :     AbstractNoArgsTypeRef
expectedPrefix       = reflect.runtime.universe.type                                                                                                :          UniqueSingleType
testedBinder         = value p3                                                                                                                     :                TermSymbol
testedTp             = reflect.runtime.universe.MethodSymbol                                                                                        :     AbstractNoArgsTypeRef
straightPrefix       = reflect.runtime.universe.type                                                                                                :          UniqueSingleType
testedTpToBounds     = reflect.runtime.universe.MethodSymbolApi with reflect.runtime.universe.TermSymbolApi with reflect.runtime.universe.SymbolApi :              RefinedType0
testedPrefix         = <notype>                                                                                                                     :                   NoType$
```

Which also reminds me of [this comment](https://github.com/lampepfl/dotty/issues/2156#issuecomment-290701098) in the Dotty repo:

> this makes it sound like there may not need more then one outer-accessor. Here's a counter-example
> 
> ```scala
> class A{ class B }
> class C extends A {self => 
>   class D {
>      final class E extends self.B  // needs two outer-accessors to support all patter-matchings
>   }
> }
> ```

Is that the fix then - to consider multiple prefixes?  The prefix of `sym.Name` is `sym` which isn't `g` but mapping to bounds `g.Name`'s prefix _is_ `g`, while the prefix of `universe.MethodSymbol` is `universe` which is what is expected.  Is that enough?  Is that unsafe?  🤷🏼 🤔 💭

Reply:
Yeah I don't know if mapping to the bounds is safe. Who says extending a bound from another prefix is not allowed?

Reply:
Oh, yes, you can do pretty nasty stuff with prefixes in extends clauses. For example:
https://github.com/scala-js/scala-js/blob/465a7645f06325ff9785bb2529c20ebd48ca3df8/test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/NestedJSClassTest.scala#L660-L691